### PR TITLE
[PT-D] Enable TP for transformer based model

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123

### CODE REVIEW

File: metaseq/launcher/opt_baselines.py [hunk 1]
```diff
@@ -97,6 +96,8 @@ def get_grid(args):
 
     total_gpus = (args.num_gpus * args.num_nodes) // size.model_parallel
     ddp_bsz = (size.batch_size // total_gpus) // SEQ_LEN
+    if args.tp_enabled:
+        ddp_bsz //= 32
```
Comment (pritamdamania87) [inline metaseq/launcher/opt_baselines.py:100]: Why do we need to do this? We shouldn't add workarounds like this to fix the OOM but actually figure out the root cause of the OOM and fix it.

File: metaseq/modules/layer_norm.py [hunk 2]
```diff
@@ -94,6 +115,13 @@ def get_statistics_from_all_workers(stat, rank, world_size):
         return buffer
 
     def forward(self, hidden_states):
+        if isinstance(hidden_states, ShardedTensor):
+            return torch.nn.functional.layer_norm(
+                hidden_states,
+                self.normalized_shape,
```
Comment (pritamdamania87) [inline metaseq/modules/layer_norm.py:121]: I'm assuming this layer norm doesn't incur any comms?

File: metaseq/modules/multihead_attention.py [hunk 3]
```diff
@@ -255,25 +258,24 @@ def forward(
...
-            q.contiguous()
-            .view(tgt_len, bsz * self.num_heads, self.head_dim)
-            .transpose(0, 1)
-        )
+        head_dim = self.head_dim
+        bsz_heads_dim = bsz * self.num_heads
+        if isinstance(self.q_proj.weight, ShardedTensor):
+            sharding_spec = self.q_proj.weight.sharding_spec()
+            world_size = len(sharding_spec.placements)
+            tgt_len *= world_size
```
Comment (pritamdamania87) [inline metaseq/modules/multihead_attention.py:266]: Could you explain the changes here?

File: metaseq/modules/multihead_attention.py [hunk 4]
```diff
@@ -312,6 +314,24 @@ def forward(
...
+                q.local_tensor(), sharding_spec, st_size
+            )
+            if k is not None:
+                st_size = (bsz * self.num_heads, k.size(1), self.head_dim)
+                k = ShardedTensor._init_from_local_tensor(
+                    k.local_tensor(), sharding_spec, st_size
+                )
+            if v is not None:
+                st_size = (bsz * self.num_heads, v.size(1), self.head_dim)
+                v = ShardedTensor._init_from_local_tensor(
```
Comment (pritamdamania87) [inline metaseq/modules/multihead_attention.py:332]: Why do we need to create ShardedTensors here? Doesn't `shard_module` appropriately shard multihead attention?

File: metaseq/modules/multihead_attention.py [hunk 5]
```diff
@@ -345,6 +365,17 @@ def forward(
...
+            zero_mask = torch.ones(
+                *attn_weights.size(), device=attn_weights.local_tensor().device
+            )
+            row, col = attn_weights.size(1), attn_weights.size(2)
+            row //= world_size
+            col //= world_size
+            for i in range(world_size):
+                zero_mask[:, row * i : row * (i + 1), col * i : col * (i + 1)] = 0
+            zero_mask = zero_mask.type(torch.BoolTensor).cuda()
+            attn_weights = attn_weights.masked_fill(zero_mask, 0.0)
```
Comment (pritamdamania87) [inline metaseq/modules/multihead_attention.py:378]: Two questions:

1) Why do we need to do this since the original code doesn't have this logic?
2) `zero_mask` is a huge tensor and we are only using part of it for masked_fill, so maybe we can allocate only that much memory?

File: metaseq/modules/multihead_attention.py [hunk 6]
```diff
@@ -357,7 +388,16 @@ def forward(
...
             if self.onnx_trace:
                 attn_mask = attn_mask.repeat(attn_weights.size(0), 1, 1)
-            attn_weights += attn_mask
+            if isinstance(attn_weights, ShardedTensor):
+                attn_mask = attn_mask.repeat(1, 8, 8)
+                attn_local = attn_weights.local_tensor()
+                attn_local += attn_mask
+                attn_weights = ShardedTensor._init_from_local_tensor(
+                    attn_local, attn_weights.sharding_spec(), attn_weights.size()
+                )
```
Comment (pritamdamania87) [inline metaseq/modules/multihead_attention.py:397]: Why do we need this change?

File: metaseq/modules/multihead_attention.py [hunk 7]
```diff
@@ -385,7 +431,20 @@ def forward(
...
+                sharding_spec.dim = -1
+                st_size = (
+                    attn.size(0),
+                    attn.size(1) // world_size,
+                    attn.size(2) * world_size,
+                )
+                attn = ShardedTensor._init_from_local_tensor(
+                    attn.local_tensor(), sharding_spec, st_size
+                )
+            attn = attn.view(tgt_len, bsz, embed_dim)
```
Comment (pritamdamania87) [inline metaseq/modules/multihead_attention.py:446]: Hmm so why can't our `view` op directly deal with this?

File: metaseq/modules/transformer_layer.py [hunk 8]
```diff
@@ -58,7 +59,14 @@ def _ffn(x, fc1, activation_fn, fc2, dropout_module, ffn_ln=None):
...
         if ffn_ln is not None:
-            x = ffn_ln(x)
+            if isinstance(x, ShardedTensor):
+                x = ffn_ln(x)
+            else:
+                m_list = x.chunk(8, dim=-1)
+                x = []
+                for m in m_list:
+                    x.append(ffn_ln(m))
+                x = torch.cat(x, dim=-1)
```
Comment (pritamdamania87) [inline metaseq/modules/transformer_layer.py:69]: why do we need this block?

File: metaseq/modules/multihead_attention.py [hunk 9]
```diff
@@ -175,10 +176,12 @@ def forward(
 
         if (
```
Comment (fduwjj) [inline metaseq/modules/multihead_attention.py:177]: Follow up: to ensure which branch the prod code is entering.

File: metaseq/modules/multihead_attention.py [hunk 10]
```diff
@@ -255,25 +258,24 @@ def forward(
...
+            if key_padding_mask is not None:
+                key_padding_mask = key_padding_mask.repeat(1, world_size)
+        q = q.contiguous().view(tgt_len, bsz_heads_dim, head_dim).transpose(0, 1)
         if k is not None:
-            k = (
-                k.contiguous()
-                .view(-1, bsz * self.num_heads, self.head_dim)
-                .transpose(0, 1)
-            )
+            k = k.contiguous().view(-1, bsz_heads_dim, head_dim).transpose(0, 1)
```
Comment (fduwjj) [inline metaseq/modules/multihead_attention.py:274]: let's create a new st here.

File: metaseq/modules/multihead_attention.py [hunk 11]
```diff
@@ -312,6 +314,24 @@ def forward(
             # In this branch incremental_state is never None
             assert incremental_state is not None
             incremental_state = self._set_input_buffer(incremental_state, saved_state)
+
+        if isinstance(q, ShardedTensor):
```
Comment (fduwjj) [inline metaseq/modules/multihead_attention.py:318]: Let's add comments here.

File: metaseq/modules/multihead_attention.py [hunk 12]
```diff
@@ -357,7 +388,16 @@ def forward(
             attn_mask = attn_mask.unsqueeze(0)
             if self.onnx_trace:
                 attn_mask = attn_mask.repeat(attn_weights.size(0), 1, 1)
-            attn_weights += attn_mask
+            if isinstance(attn_weights, ShardedTensor):
```
Comment (fduwjj) [inline metaseq/modules/multihead_attention.py:391]: Add comment for this walk-around.

File: metaseq/modules/transformer_layer.py [hunk 13]
```diff
@@ -58,7 +59,14 @@ def _ffn(x, fc1, activation_fn, fc2, dropout_module, ffn_ln=None):
         x = fc1(x)
         x = activation_fn(x)
         if ffn_ln is not None:
-            x = ffn_ln(x)
+            if isinstance(x, ShardedTensor):
+                x = ffn_ln(x)
+            else:
```
Comment (fduwjj) [inline metaseq/modules/transformer_layer.py:64]: Let's remove it

File: metaseq/tasks/base_task.py [hunk 14]
```diff
@@ -405,12 +406,22 @@ def train_step(
...
         with torch.autograd.profiler.record_function("forward"):
             loss, sample_size, logging_output = criterion(model, sample)
+        print("logging output inside task ", logging_output, file=sys.stderr)
+        print("ignore flag: ", ignore_grad, file=sys.stderr)
+        print("model : ", model, file=sys.stderr)
+        print("flag in criterion ", model.has_full_params, file=sys.stderr)
+        #loss.backward()
         if ignore_grad:
             loss *= 0
+        loss /= 1000
```
Comment (fduwjj) [inline metaseq/tasks/base_task.py:419]: maybe let's remove this.

File: metaseq/trainer.py [hunk 15]
```diff
@@ -847,14 +861,8 @@ def maybe_no_sync():
...
                 "loss_scale",
-                self.optimizer.scaler.loss_scale,
-                priority=700,
-                round=4,
-                weight=0,
-            )
-            metrics.log_scalar(
-                "scale_window",
-                self.optimizer.scaler.scale_window,
+                #self.optimizer.scaler.loss_scale,
```
Comment (fduwjj) [inline metaseq/trainer.py:864]: Let's recover the deleted code here.

File: metaseq_cli/train.py [hunk 16]
```diff
@@ -93,18 +142,24 @@ def main(cfg: DictConfig) -> None:
     assert cfg.criterion, "Please specify criterion to train a model"
 
     # Build model and criterion
+    built_model = task.build_model(cfg.model)
+    if cfg.distributed_training.tp_enabled:
```
Comment (fduwjj) [inline metaseq_cli/train.py:146]: for 175B model, we need to use meta tensor to initialize the model. Not for June

File: metaseq/modules/layer_norm.py (hunk 2 above)
Comment (fduwjj) [reply metaseq/modules/layer_norm.py:121]: No this does need comms, because it's "synced."

File: metaseq/modules/multihead_attention.py (hunk 11 above)
Comment (fduwjj) [reply metaseq/modules/multihead_attention.py:318]: tgt_len, bsz, embed_dim = query.size()

after linear

tgt_len, bsz, embed_dim // paralle_size - col-wise sharding original q size. sharded by dim 2

tgt_len, bsz * self.num_heads_partition, self.head_dim implicitly change the sharding dim to 1.

File: metaseq/modules/multihead_attention.py (hunk 5 above)
Comment (fduwjj) [reply metaseq/modules/multihead_attention.py:378]: Because we now gather input from different ranks and doing bmm will create interaction between weights between different ranks which are not needed. (They have no meaning, IMHO)

Comment (fduwjj) [reply metaseq/modules/multihead_attention.py:378]: And I agree, we can have a more efficient way to do this.

File: metaseq/modules/multihead_attention.py (hunk 9 above)
Comment (fduwjj) [reply metaseq/modules/multihead_attention.py:177]: Looks like megatron also skip this condition.

File: metaseq/modules/multihead_attention.py (hunk 4 above)
Comment (fduwjj) [reply metaseq/modules/multihead_attention.py:332]: Megatron implicitly change the sharding spec during view.

File: metaseq/modules/multihead_attention.py (hunk 7 above)
Comment (fduwjj) [reply metaseq/modules/multihead_attention.py:446]: Because View does not change sharding spec automatically.

**Comment by fduwjj** [comment]
Created a fork https://github.com/fduwjj/metaseq_ptd to merge all local changes first and faster iteration.
